nemo/collections/nlp/models/information_retrieval/bert_embedding_model.py

```diff
@@ -0,0 +1,141 @@
+
+except (ImportError, ModuleNotFoundError):
+
+    ModelParallelConfig = ApexGuardDefaults
+
+    HAVE_MEGATRON_CORE = False
+
+
+class BertEmbeddingHead(nn.Module):
+    """Performs mean pooling on the token embeddings.
+    """
+
+    def __init__(
+        self, word_embedding_dimension: int, pooling_mode_mean_tokens: bool = True,
+    ):
+        super(BertEmbeddingHead, self).__init__()
+
+        self.config_keys = [
+            "word_embedding_dimension",
+            "pooling_mode_mean_tokens",
+        ]
+        self.word_embedding_dimension = word_embedding_dimension
+        self.pooling_mode_mean_tokens = pooling_mode_mean_tokens
+
+    def forward(self, token_embeddings: Tensor, attention_mask: Tensor):
+
+        token_embeddings = token_embeddings.permute(1, 0, 2)
+        input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
+        sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)
+
+        sum_mask = input_mask_expanded.sum(1)
+
+        sum_mask = torch.clamp(sum_mask, min=1e-9)
+
+        output_vector = sum_embeddings / sum_mask
+
+        output_vector = F.normalize(output_vector, p=2, dim=1)
+
+        return output_vector
+
+    def __repr__(self):
+        return "Pooling({}) and Normalize".format(self.get_config_dict())
+
+    def get_config_dict(self):
+        return {key: self.__dict__[key] for key in self.config_keys}
+
+
+class MCoreBertEmbeddingModel(MCoreBertModelWrapper):
+    def __init__(self, *args, **kwargs):
+
+        super(MCoreBertEmbeddingModel, self).__init__(*args, **kwargs)
+        # Changing the default settings of the original Bert model to make it compatible with the embedding model.
+        self.post_process = False
+        self.binary_head = None
+        self.lm_head = None
+        self.output_layer = None
+        self.encoder.final_layernorm = None
+        self.encoder.post_process = False
+        self.embedding_head = BertEmbeddingHead(
+            word_embedding_dimension=self.config.hidden_size, pooling_mode_mean_tokens=True,
+        )
+
+    def forward(
+        self,
+        input_ids: Tensor,
+        attention_mask: Tensor,
+        tokentype_ids: Tensor = None,
+        lm_labels: Tensor = None,
+        inference_params=None,
+    ):
+        """Forward function of BERT model
+
+        Forward function of the BERT Model This function passes the input tensors
+        through the embedding layer, and then the encoder and finally into the post
+        processing layer (optional).
+
+        It either returns the Loss values if labels are given  or the final hidden units
+        """
+        hidden_states = super(MCoreBertEmbeddingModel, self).forward(
+            input_ids, attention_mask, tokentype_ids, lm_labels, inference_params
+        )
+        embeddings_out = self.embedding_head(hidden_states, attention_mask)
+        return embeddings_out
+
+
+class BertEmbeddingModel(BertModel):
```

Comment:
Can you rename bot of these to NeMoBertEmbeddingModel(NeMoBertModel) (or import it as this up to you)
and add a warning saying this will be deprecated mid 2024 (Also make a note here)

Reply:
added warnings.warn("NeMoBertModel will be deprecated mid 2024. Use MCoreBertModelWrapper instead.", DeprecationWarning)